### ISSUE TITLE
Debugging Python C sources with PythonView

### ISSUE THREAD

**Issue by amos402** [Microsoft/PTVS#2657]
For debugging `.pyd` in PTVS， when tracking in the C sources it's easy to see the Python objects' information. Is there any way to enable this feature when I just debugging the Python core sources?

**Comment by zooba** [comment]
Yes, though there are a few little tricks.

The first is you need to set a registry key to enable some secret options. If you create `HKEY_CURRENT_USER\Software\Microsoft\PythonTools\Debugger` and add a DWORD value `PythonDeveloper` with the value 1 (assuming you're on the latest VS 2017 - if not, let me know and I'll get you the correct key for whatever version you have).

Next, you need to attach the debugger. This is a little tricky right now (we have some work to make it easier), but the best way to do it is to start python.exe running and then Debug/Attach to it. When you do this, you'll want to customize the debug engines to include "Native" and "Python".

Once that is done, you'll get some extra right-click menus where the "Show Python view" menu normally is. The "Show Native Python Frames" option is the one you want - normally we don't let you step through `pythonXY.dll`, but this option will enable that.

(The future work we have to do for the second step is make our debugger an option for C++ projects. We know how to do it, just haven't written or released it yet.)

(+ @int19h to correct anything I got wrong)

**Comment by int19h** [comment]
Slight correction: the "Show Native Python Frames" will show native frames from python*.dll and _ctypes.pyd in the stack, so you can see the interpreter state; but it doesn't enable native stepping.

There's a separate option there called "Use Python Stepping" - which is enabled by default - which makes stepping inside those modules (regardless of whether you're viewing the exposed native frame, or the synthesized Python frame) behave semantically as a Python step, i.e. run the debuggee until the next Python line or breakpoint. Unchecking that option will disable this Python semantics, at which point all steps are handled by the native debugger.

Again, it doesn't matter if you're viewing the C source or the Python source - when this option is unchecked, and you're e.g. at a Python breakpoint (and looking at the .py file in the editor), stepping will step through the C code for the main Python interpreter loop.

**Comment by amos402** [comment]
> create HKEY_CURRENT_USER\Software\Microsoft\PythonTools\Debugger and add a DWORD value PythonDeveloper with the value 1
I have done with these steps, but I can't find the `Show Native Python Frames` option anywhere, the `Use Python Stepping` also.

By the way, I usually use the watch window change the object pointer to the expect type pointer to see the structure. This is a troublesome way, Any good idea for seeing the variable just like `libpython` in GDB?

**Comment by zooba** [comment]
> Any good idea for seeing the variable...

Yep, we're trying to get you using it right now :)

Did you catch the bit about using Debug/Attach rather than F5? Right now, F5 is going to use the native-only debugger, so you need to Debug/Attach to get native and Python together.

**Comment by amos402** [comment]
Yes, `Attach`, not `Run`.
Besides, I still mind where can I find `Show Native Python Frames` and `Use Python Stepping`.😅

And I found there are 3 situation using attach.

Registry I seted:
![_20170610_035057](https://user-images.githubusercontent.com/3406561/26992187-0ebd87ae-4d90-11e7-9332-23522ab591fa.png)

If I set the attach option to `Python Code` and `Native Code`, It will only break in `main` funtion, but my break point was on `bltinmodule.c`.
![_20170610_032853](https://user-images.githubusercontent.com/3406561/26991460-f6fc7420-4d8c-11e7-8ea9-6b6577572a35.png)
**本机 代码 => Native Code**, I haven't install English pack, I don't know the original words exactly 😹

![_20170610_032425](https://user-images.githubusercontent.com/3406561/26991292-60d1fefc-4d8c-11e7-8961-a306f1ece55a.png)

If I only choose `Native Code`, it would break on my break point I setted, but no PythonView
![_20170610_034413](https://user-images.githubusercontent.com/3406561/26991968-188cdee8-4d8f-11e7-8acf-ce91f1831ccb.png)

If I use the auto option or only `Python Code`, It can only break in py source.

**Comment by int19h** [comment]
Those two options show up in the context menu - right-click on e.g. the Locals window, and you should see the Python item there. The options are underneath that.

Your first screenshot shows the correct behavior with the options set to default values - it does actually hit your breakpoint, but because it is inside python.dll, the call stack doesn't show that, since it hides all python.dll frames (and synthesizes Python frames out of them). Once you enable "Show Native Python Frames", you will see it as it actually is under the hood.

**Comment by amos402** [comment]
I found `Show Python View Nodes` on locals window context menu, but I still can't find `Show Native Python Frames` and `Use Python Stepping` normally. 
Finally I found them in customize menu and add them to the menu temporary. Is any other way to get them genteelly? (+o+)
I checked `Show Native Python Frames`(**显示本机 Python 框架** in screenshot) it shows up the frams in python.dll. 

But even I checked `Show Python View Nodes`, I still can't see the python view.
![_20170610_141840](https://user-images.githubusercontent.com/3406561/27000556-cd4d1e6c-4de7-11e7-8c4f-dbd07d88b291.png)

And with `Use Python &Stepping`, I can't step into a function call using F11. If I uncheck it, it would go straight just like F10
this is a pre frame:
![_20170610_143634](https://user-images.githubusercontent.com/3406561/27000678-3843142c-4dea-11e7-9dc3-63bdf8629fe6.png)

after F11:
![_20170610_143021](https://user-images.githubusercontent.com/3406561/27000643-664f2712-4de9-11e7-81db-8813b3528537.png)

**Comment by int19h** [comment]
The registry key is really what's supposed to make them show up... I wonder if something regressed in our command placement. If you can add them manually, I can't think of any other reason. I'll look into it.

Can you see the Python view nodes in native code outside of python.dll? 

And I'm not sure what's going on with stepping there. It doesn't even look like F10, since it didn't end up on the next line. It actually looks more like F11 was handled by the Python stepping logic, which treated it as stepping into a native (as in, outside of python.dll) function call - which would be why it stopped where it is on the stack. But it shouldn't be trying to handle F11 at all here.

To clarify, does F10 work for you as expected - i.e. does it step over a single line of *native* code?

**Comment by amos402** [comment]
It can hit the break point on pyd source, but still have no python view.

>To clarify, does F10 work for you as expected - i.e. does it step over a single line of native code?  
No, after F10 it just go to PTVS's frame
![_20170611_012243](https://user-images.githubusercontent.com/3406561/27004899-c32deb84-4e44-11e7-8ecf-ae60df61ad9d.png)

**Comment by int19h** [comment]
So you don't see any observed difference between "Use Python Stepping" being checked or unchecked?

**Comment by amos402** [comment]
> So you don't see any observed difference between "Use Python Stepping" being checked or unchecked?  

Yes, they didn't work correctly as I expected.

I just try these again on VS2015(above results are base on VS2017) with PTVS 2.2.5, on VS2015, it can show the Python view, and the context menu on local window has 4 options: `Show C++ View Nodes`, `Show Python View Nodes`, `Show Native Python Frames` `Use Python Stepping`. But F10 and F11 still not work correctly.

So...it means they are new bugs on PTVS3.0(released on VS15.2).

**Comment by zooba** [comment]
I'm not 100% sure whether this is still an issue (I saw it recently but it ended up being unrelated to this, as far as I can tell). We should check it again as soon as possible.

**Comment by int19h** [comment]
I can't repro stepping issues on our most current bits.

The issue with Python View is because we're registering our .natvis file incorrectly. This is from error log with NatVis warnings enabled:
```
Natvis: Fatal error: Unable to open natvis file 'C:\USERS\INT19\APPDATA\LOCAL\MICROSOFT\VISUALSTUDIO\15.0_B6255AABEXP\EXTENSIONS\MICROSOFT CORPORATION\PYTHON\3.0\PythonDkm.natvis'.
```
The path where it's actually installed is `...\3.0\Visualizers\PythonDkm.natvis`, but our .vsixmanifest doesn't include that subpath. Need to fix either the manifest or the install location.

**Comment by int19h** [comment]
I'll spin off a separate bug for Python View:
https://github.com/Microsoft/PTVS/issues/3018

**Comment by huguesv** [comment]
It looks like the bug has been fixed.
